shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm

```diff
@@ -153,24 +179,71 @@ - (void)netService:(NSNetService*)sender didNotPublish:(NSDictionary*)errorDict
                     "network settings and relaunch the application.";
 }
 
-#if TARGET_IPHONE_SIMULATOR
-static void DNSSD_API registrationCallback(DNSServiceRef sdRef,
-                                           DNSServiceFlags flags,
-                                           DNSServiceErrorType errorCode,
-                                           const char* name,
-                                           const char* regType,
-                                           const char* domain,
-                                           void* context) {
-  if (errorCode == kDNSServiceErr_NoError) {
-    FML_DLOG(INFO) << "FlutterObservatoryPublisher is ready!";
+@end
+
+@implementation FlutterObservatoryPublisher {
+  flutter::DartServiceIsolate::CallbackHandle _callbackHandle;
+  std::unique_ptr<fml::WeakPtrFactory<FlutterObservatoryPublisher>> _weakFactory;
+}
+
+- (NSURL*)url {
+  return [_delegate.get().url copy];
+}
+
+- (instancetype)init {
+  self = [super init];
+  NSAssert(self, @"Super must not return null on init.");
+
+  if (@available(iOS 9.3, *)) {
+    _delegate.reset([[ObservatoryDNSServiceDelegate alloc] initWithOwner:self]);
   } else {
-    FML_LOG(ERROR) << "Could not register as server for FlutterObservatoryPublisher. Check your "
-                      "network settings and relaunch the application.";
+    _delegate.reset([[ObservatoryNSNetServiceDelegate alloc] initWithOwner:self]);
+  }
+  _weakFactory = std::make_unique<fml::WeakPtrFactory<FlutterObservatoryPublisher>>(self);
+
+  fml::MessageLoop::EnsureInitializedForCurrentThread();
+
+  _callbackHandle = flutter::DartServiceIsolate::AddServerStatusCallback(
+      [weak = _weakFactory->GetWeakPtr(),
+       runner = fml::MessageLoop::GetCurrent().GetTaskRunner()](const std::string& uri) {
+        if (!uri.empty()) {
+          runner->PostTask([weak, uri]() {
+            if (weak) {
+              [[weak.get() delegate]
+                  publishServiceProtocolPort:[NSString stringWithUTF8String:uri.c_str()]];
+            }
+          });
+        }
+      });
+
+  return self;
+}
+
+- (NSString*)serviceName {
+  return [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleIdentifier"];
+}
+
+- (NSData*)createTxtData:(NSURL*)url {
+  // Check to see if there's an authentication code. If there is, we'll provide
+  // it as a txt record so flutter tools can establish a connection.
+  auto path = std::string{[[url path] UTF8String]};
+  if (!path.empty()) {
+    // Remove leading "/"
+    path = path.substr(1);
```

Comment:
Do we need to use std::string here?  Since it's Obj-C -> Obj-C it would be nice to not need to translate between unless there's a good reason.

```objective-c
NSString *path = [url.path stringByReplacingOccurrencesOfString:@"/" withString:@"" options:NSLiteralSearch range:NSMakeRange(0, 1)];
// OR if you don't want any trailing / either
NSString *path = [url.path stringByTrimmingCharactersInSet:[NSCharacterSet characterSetWithCharactersInString:@"/"]];
NSData *data = [path dataUsingEncoding:NSUTF8StringEncoding];
NSDictionary<NSString*,NSData*> *authCode = @{
  @"authCode" : data
};
return [NSNetService dataFromTXTRecordDictionary:authCode];
```

Reply:
Done - slightly different than your suggestion but yeah, that was legacy code from something else.